File: backend/src/taxonomy_builder/schemas/publishing.py
"""Pydantic schemas for the publishing workflow API."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from taxonomy_builder.schemas.snapshot import DiffResult, ValidationResult


class ContentSummary(BaseModel):
    """Summary of what a project contains."""

    schemes: int = 0
    concepts: int = 0
    properties: int = 0
    classes: int = 0


class PublishPreview(BaseModel):
    """Response from the preview endpoint: validation + diff + content summary."""

    validation: ValidationResult
    diff: DiffResult | None = None
    content_summary: ContentSummary
    suggested_version: str | None = None
    suggested_pre_release_version: str | None = None
    latest_version: str | None = None
    latest_pre_release_version: str | None = None


VERSION_PATTERN = r"^\d+(\.\d+)+(-pre\d+)?$"


PRE_RELEASE_SUFFIX = r"-pre\d+$"


class PublishRequest(BaseModel):
    """Request body for publishing a version."""

    version: str = Field(pattern=VERSION_PATTERN)
    title: str
    notes: str | None = None
    pre_release: bool = False

    @model_validator(mode="after")
    def version_matches_pre_release_flag(self) -> "PublishRequest":
        import re

        has_suffix = bool(re.search(PRE_RELEASE_SUFFIX, self.version))
        if self.pre_release and not has_suffix:
            raise ValueError("pre-release versions must have a -preN suffix")
        if not self.pre_release and has_suffix:
            raise ValueError("release versions must not have a -preN suffix")
        return self


class PublishedVersionRead(BaseModel):
    """Read model for a published version."""

    model_config = {"from_attributes": True}

    id: UUID
    project_id: UUID
    version: str
    title: str
    notes: str | None = None
    finalized: bool
    published_at: datetime
    publisher: str | None = None
    latest: bool = False

    previous_version_id: UUID | None = None

File: frontend/src/api/publishing.ts
import { api } from "./client";

// ============ Types ============

export interface ValidationError {
  code: string;
  message: string;
  entity_type: string | null;
  entity_id: string | null;
  entity_label: string | null;
}

export interface ValidationResult {
  valid: boolean;
  errors: ValidationError[];
}

export interface DiffItem {
  id: string | null;
  uri: string | null;
  label: string;
  entity_type: string;
}

export interface FieldChange {
  field: string;
  old: string | null;
  new: string | null;
}

export interface ModifiedItem {
  id: string;
  label: string;
  entity_type: string;
  changes: FieldChange[];
}

export interface DiffResult {
  added: DiffItem[];
  modified: ModifiedItem[];
  removed: DiffItem[];
}

export interface ContentSummary {
  schemes: number;
  concepts: number;
  properties: number;
}

export interface PublishPreview {
  validation: ValidationResult;
  diff: DiffResult | null;
  content_summary: ContentSummary;
  suggested_version: string | null;
  suggested_pre_release_version: string | null;
  latest_version: string | null;
  latest_pre_release_version: string | null;
}

export interface PublishRequest {
  version: string;
  title: string;
  notes?: string | null;
  pre_release?: boolean;
}

export interface PublishedVersionRead {
  id: string;
  project_id: string;
  version: string;
  title: string;
  notes: string | null;
  finalized: boolean;
  published_at: string | null;
  publisher: string | null;
  latest: boolean;
  previous_version_id: string | null;
}

// ============ API ============

export const publishingApi = {
  getPreview: (projectId: string) =>
    api.get<PublishPreview>(`/projects/${projectId}/publish/preview`),

  publish: (projectId: string, data: PublishRequest) =>
    api.post<PublishedVersionRead>(`/projects/${projectId}/publish`, data),

  listVersions: (projectId: string) =>
    api.get<PublishedVersionRead[]>(`/projects/${projectId}/versions`),
};
